scripts/pylint_extensions.py

```diff
@@ -888,13 +894,80 @@ def process_module(self, node):
             node: astroid.scoped_nodes.Function. Node to access module content.
         """
 
-        file_content = node.stream().readlines()
-        file_length = len(file_content)
+        with node.stream() as stream:
+            file_content = stream.readlines()
+            file_length = len(file_content)
 
-        if file_length == 1 and len(file_content[0]) == 1:
-            self.add_message('only-one-character', line=file_length)
-        if file_length >= 2 and not re.search(r'[^\n]\n', file_content[-1]):
-            self.add_message('newline-at-eof', line=file_length)
+            if file_length == 1 and len(file_content[0]) == 1:
+                self.add_message('only-one-character', line=file_length)
+            if file_length >= 2 and not re.search(r'[^\n]\n',
+                                                  file_content[-1]):
+                self.add_message('newline-at-eof', line=file_length)
+            for (line_num, line) in enumerate(file_content):
+                try:
+                    # Checks newlines preceding top-level function.
+                    if (line.startswith('def ') and
```

Comment:
@YashJipkate, I am not super happy about the way we are checking this...reading the whole file and then checking for all edge cases does not seem to be the right way to do this. There can be an edge case in the future not covered by this. Can you please look into any other alternatives or optimize this approach to ensure so that it's foolproof?Thanks!

Reply:
@apb7 I was also not fully satisfied with this approach owing to its heavy repetition of code and as you said - vulnerability.
But these are really the only cases possible in our codebase right now. About your concerns about the future, I tried some options like using the IAstroidChecker instead of the IRawChecker that accurately catches the functions, but its problem being that it does not concern itself with what precedes it, and the only way to do this was to parse the whole file using IRawChecker, but since IRawChecker cant identify functions/class definitions, hence the hard check.
I'll try to find out some way once my exams are over :P but until then if you could think of something. it would be really helpful.
Thanks!